sklearn/compose/_column_transformer.py

```diff
@@ -409,15 +415,30 @@ def fit_transform(self, X, y=None):
         result = self._fit_transform(X, y, _fit_transform_one)
 
         if not result:
+            self._update_fitted_transformers([])
             # All transformers are None
             return np.zeros((X.shape[0], 0))
 
         Xs, transformers = zip(*result)
 
+        # determine if concatenated output will be sparse or not
+        if all(sparse.issparse(X) for X in Xs):
+            self._sparse = True
+        elif any(sparse.issparse(X) for X in Xs):
+            nnz = sum(X.nnz if sparse.issparse(X) else X.size for X in Xs)
+            total = sum(X.shape[0] * X.shape[1] if sparse.issparse(X)
+                        else X.size for X in Xs)
+            if (nnz / total) < self.sparse_threshold:
```

Comment:
Maybe it's worth storing `density_` so that users can get familiar with this calculation and tweak it.

Reply:
That would then mean adding two attributes? (as @ogrisel also mentioned adding `self.sparse_ouptut_`)

Reply:
Ah. I'd consider `density_` as an alternative to `sparse_output_`. It would allow us to more clearly document what determines sparse output, which would still just be a matter of numerical comparison.

Reply:
On the one hand, `density_` is more general as `sparse_output_`, as one can (indirectly) get the same by doing `density_ < sparse_threshold`. 
But, you have the corner cases of no-sparse / all-sparse were this is not fully clear what the result of `density_` would be. 

Eg for all-sparse: currently I don't calculate the density but always set `sparse_output_` to True (in the idea that if all outputs of the transformers are sparse, the combined output will also always be sparse)